src/mon/Monitor.cc

```diff
@@ -2832,6 +2832,14 @@ void Monitor::do_health_to_clog(bool force)
       level == health_status_cache.overall)
     return;
   clog->health(level) << "overall " << summary;
+  if (level != HEALTH_OK) {
+    Formatter *f = Formatter::create("json-pretty");
+    health_status_t level = healthmon()->get_health_status(true, f, nullptr);
+    stringstream ss;
+    f->flush(ss);
+    clog->health(level) << "Health detail:\n" << ss.str();
```

Comment:
is the clog supposed to be parsed by a json parser?

Reply:
Yes. If we don't parse it in json format then log entry will look like [914729](http://pastebin.test.redhat.com/914729)

Reply:
could you paste the log in a public-accessible place?

Reply:
[cluster log without json formatted](https://paste.centos.org/view/13dbdcf9)

Reply:
@pdvian i assume the cluster log is consumed by human being most of the time. could you help me understand where the json parser kick in in a typical use case?

also, per the description of the tracker ticket

> This will help us spot things like obvious network issues which can lead to racks/hosts down in a cluster. Also gives a historic view of a cluster's health.

@neha-ojha @joshd

if we have rack/host down issues, would `ceph -s` show it in a brief form?

also i am not sure to dump the health report to clog every 10 min is the best way to keep track of the history. by default, we keep 500 epochs. if there are some more clog messages apart from the health report in an unhealthy cluster, the clog will trim the history very soon.

also, i think the log of monitor is good enough if we want to have the history.

Reply:
> @pdvian i assume the cluster log is consumed by human being most of the time. could you help me understand where the json parser kick in in a typical use case?
> 
> also, per the description of the tracker ticket
> 
> > This will help us spot things like obvious network issues which can lead to racks/hosts down in a cluster. Also gives a historic view of a cluster's health.
> 
> @neha-ojha @jdurgin
> 
> if we have rack/host down issues, would `ceph -s` show it in a brief form?

It doesn't show which osds/hosts/racks are affected.

> also i am not sure to dump the health report to clog every 10 min is the best way to keep track of the history. by default, we keep 500 epochs. if there are some more clog messages apart from the health report in an unhealthy cluster, the clog will trim the history very soon.

clog goes to disk in the cluster log file. That it's stored in the monitor db at all is another bug imo.

> also, i think the log of monitor is good enough if we want to have the history.

The cluster log is for messages like this that are meant to be understood by humans.

Adding something to only the daemon logs places too much overhead on administrators - anything we want them to know about should go in the cluster log.

Reply:
@tchaikov you have a good point about json - it would be easier to read if it was plain format, and we logged one line at a time

Reply:
@tchaikov updated PR as per above discussion. Kindly review it.